lib/graphiti/resource_proxy.rb

```diff
@@ -154,7 +154,7 @@ def debug_requested?
     private
 
     def persist
-      @resource.transaction do
+      result, after_hooks = @resource.transaction do
```

Comment:
This is kinda odd, and took me a second to understand the different interfaces of `Hooks.run` vs returning`after_hooks`. That said, I get it now and think it's probably the best we can do. A comment here or at the top of `hooks.rb` would be helpful six months from now, I think.

Ideally, we move validations to this same "record a proc for later playback" pattern, which I think will simplify things and cause us to make `Util::Hooks` a more first-class object. IOW I think a private method was/is fine, but this is now smelling like the need for a separate class holding this logic. Not necessarily required for this commit but the direction I'd like to go down.

Reply:
I've updated `Util::Hooks` to be a more lifecycle aware `Util::TransactionHooksRecorder`.  Instead of returning an implicit array, results are now expected to be returned as a structured `{results:, after_commit_hooks:}` hash.  This isn't quite as clean as I'd like since the `Resource#transaction` method bypasses this structuring when validation errors happen and I need to repeat in the rescue block, and the destroy path doesn't use the hook recorder at all, so some destructuring was needed there, but I think in general this is a lot more readable.